### PR TITLE
Fix offenses to non-production code

### PR DESCRIPTION
**Pull request by tagliala**
- Lint/AmbiguousBlockAssociation
- Lint/AmbiguousOperatorPrecedence
- Lint/EmptyBlock
- Lint/RedundantSplatExpansion
- Lint/RedundantStringCoercion
- Lint/SymbolConversion
- Lint/UnusedBlockArgument
- Style/BlockDelimiters
- Style/CommentAnnotation
- Style/EachForSimpleLoop
- Style/EmptyMethod
- Style/ExpandPathArguments
- Style/FileWrite
- Style/GlobalStdStream (UNSAFE)
- Style/HashEachMethods (UNSAFE)
- Style/NestedParenthesizedCalls
- Style/NilComparison
- Style/NumericLiteralPrefix (manually fixed)
- Style/ParallelAssignment
- Style/PreferredHashMethods (UNSAFE)
- Style/RedundantInterpolation (UNSAFE)
- Style/RedundantParentheses (UNSAFE)
- Style/RegexpLiteral
- Style/RescueStandardError
- Style/SpecialGlobalVars (UNSAFE)
- Style/SymbolProc (UNSAFE)
- Style/ZeroLengthPredicate (UNSAFE)

### Description
Please describe your pull request. Thank you for contributing! You're the best.

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [ ] I added an entry to the [changelog](../blob/master/CHANGELOG.md).

### CODE REVIEW

**Review by kiskoza — APPROVED** [review]
Hey, thanks for the PR! The code changes look good to me, but they surfaced a few tests that might not work as intended - we can address them in an other PR to keep this one clean or just resolve them in a new commit, either could work for me.

However, I would like to postpone its merge and create a clean release for v4.0 containing only the breaking changes to make it easier to upgrade for others.

File: test/workbook/worksheet/tc_sheet_protection.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ def test_initialize
   end
 
   def test_boolean_attribute_validation
-    @boolean_options.each do |key, value|
+    @boolean_options.each do |key, _value|
```
[the remark this answers is not in the document]
Comment (tagliala) [reply test/workbook/worksheet/tc_sheet_protection.rb:47]: Thanks, there were also other occurrences

**Comment by tagliala** [comment]
Just pushed 3 new commits to fix highlighted issues

File: test/workbook/worksheet/tc_sheet_protection.rb [hunk 2]
```diff
@@ -44,10 +44,9 @@ def test_initialize
   end
 
   def test_boolean_attribute_validation
-    @boolean_options.each do |key, _value|
+    @boolean_options.each do |key, value|
       assert_raise(ArgumentError, "#{key} must be boolean") { @sp.send("#{key}=".to_sym, 'A') }
-      assert_nothing_raised { @sp.send("#{key}=".to_sym, true) }
-      assert_nothing_raised { @sp.send("#{key}=".to_sym, true) }
+      assert_nothing_raised { @sp.send("#{key}=".to_sym, value) }
```
Comment (tagliala) [inline test/workbook/worksheet/tc_sheet_protection.rb:49]: was it intended? Didn't get it from blame

**Comment by tagliala** [comment]
Rebased.

Coverage stays the same `5450 / 5458 LOC (99.85%) covered`

**Comment by noniq** [comment]
Thanks! Please also add Rubocop version (1.50.2?) to the Gemfile.

**Comment by tagliala** [comment]
Hi,

I understand that is to prevent unexpected failures on CI caused by RuboCop updates

I've used this approach:

- Disable `NewCops`
- Lock Rubocop to < 2
- Lock RuboCop minitest to < 0.31.0

To lock a specific version, I should have exclude its installation from Ruby < 2.5 and that would require some extra effort (rubocop group and `BUNDLE_WITH`?). Let me know if you prefer a different approach and I'm also available to suggestions because I don't know what would be the best approach

**Comment by noniq** [comment]
Ah, good point about CI – I didn’t think about that, only about running Rubocop locally during development.

Can we simply lock the version in Gemfile to a minor version (`~> 1.50.2`)? Or does this lead to problems on CI and/or with older versions of Ruby?

**Comment by tagliala** [comment]
Conditionally restricted to RUBY_VERSION >= '2.6'
Disabled new cops
locked to patch versions

**Comment by straydogstudio** [comment]
@tagliala Are you on the slack channel?

**Comment by tagliala** [comment]
Blocked by #214

**Pull request merged**